docs(conventions): define artefact-linking label convention (#146)
open-component-model/ocm-spec: doc/01-model/06-conventions.md

Another possibility is to use dedicated [labels](./03-elements-sub.md#labels) to describe the usage scenario for dedicated artifacts. Here, the tool working on a component versions does not read a description artifact, but has to analyse the label settings of all the provided artifacts. In both cases there is a dedicated OCM specific interpretation of content provided by the component model. But while the first solution allows to describe a closed scenario in a dedicated resource, where resources from dependent component version can be described by relative resource references and multiple scenarios can be separated by multiple flavors of this resource, the label-based approach is restricted to a local component version and a single scenario. Instead of an artifact type for the description, labels with a defined [name structure](./03-elements-sub.md#labels) are required.

## Artefact-Linking Label

This section defines a label convention for expressing cross-artefact relationships.
For example, it can be used to indicate that an SBoM resource describes a specific
OCI image resource within the same component version.

### Label Name

```

> ocm.software/artefact-reference

Reviewer: ```suggestion
ocm.software/artefact-references
```
Since we are now allowing a list of references :-)

```

> The label follows the [vendor-specific label naming scheme](./07-extensions.md#label-types).

Reviewer: I think this should not be a vendor-specific label, but instead
> labels with a predefined meaning within the component model.

The label version is expressed via the separate `version` field on the label object
(e.g. `version: v1`); consumers MUST NOT treat a label with a different version as
conforming to this convention.

### Placement

The label is placed on the **derived artefact**. The resource that is related to a
subject artefact carries the label pointing back to that subject. The subject
artefact itself requires no modification.

```
Component Descriptor
├── Resource: my-image          ← subject artefact, unchanged
└── Resource: my-image-sbom     ← derived artefact, carries the label
```

When multiple derived artefacts reference the same subject (e.g. two SBoMs produced
by different tools for the same image), they MUST be told apart from each other by

> their own `extraIdentity`. The `identitySelector` in each label still points to the

Reviewer: ```suggestion
their own artefact identity. The `identitySelector` in each label still points to the
```
I'd say it is also valid to just use different names for those artefacts (we should not restrict this to the `extraIdentity` only).

    extraIdentity:
      foo: bar

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: amd64
    labels:
      - name: ocm.software/artefact-reference
        version: v1
        value:
          - identity:
              name: my-image
              version: 1.2.3
              foo: bar

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: arm64
    labels:
      - name: ocm.software/artefact-reference
        version: v1
        value:
          - identity:
              name: my-image
              version: 1.2.3
              foo: bar
```

### Lookup Algorithm

To find all artefacts related to a given subject resource:

1. Determine the identity of the subject resource: its `name`, `version`, and `extraIdentity`.
2. Iterate over all resources in the component descriptor.
3. For each resource, check whether it carries a label named `ocm.software/artefact-reference` with `version: v1`.
4. If present, apply the following matching rules against each `identity` entry in the label value:
   - `name` MUST equal the subject's `name`.
   - If `version` is set, it MUST equal the subject's `version`.

>    - Every additional key-value pair MUST be present and equal in the subject's `extraIdentity`.

Reviewer: Also, the subject's `extraIdentity` must not contain any additional fields which are not part of the `identity` entry (it should _not_ serve as a generous selector, but instead an explicit identity match).

- Any additional key-value pair is treated as an extra identity property.
  Every such entry MUST be present and equal in the subject's `extraIdentity`.
  Required when multiple resources share the same name (e.g. arch-specific image variants).

### Examples

An SBoM describing a single-variant image:

```yaml
resources:
  - name: my-image
    version: 1.2.3
    type: ociImage

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    labels:
      - name: ocm.software/artefactReference
        version: v1
        value:
          identitySelector:
            name: my-image
```

Two SBoMs referencing the same subject, told apart by their own `extraIdentity`:

```yaml
resources:
  - name: my-image
    version: 1.2.3
    type: ociImage
    extraIdentity:
      foo: bar

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: amd64
    labels:
      - name: ocm.software/artefactReference
        version: v1
        value:
          identitySelector:
            name: my-image
            version: 1.2.3
            foo: bar

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: arm64

>     labels:

Reviewer 2: ```
      - name: ocm.software/artefactReferences # check whether we want different conventions than camel case
        version: v1
        value:
          - identity:
              name: my-image
              version: 1.2.3
              foo: bar
            metadata: # once we need it

```

Reviewer 2: even for in-house labels, we might want to prefix with ocm.software

Reviewer 3: If we want to follow conventions on K8s annotations & labels, it would be good to have ocm.software as a prefix :-)

Another possibility is to use dedicated [labels](./03-elements-sub.md#labels) to describe the usage scenario for dedicated artifacts. Here, the tool working on a component versions does not read a description artifact, but has to analyse the label settings of all the provided artifacts. In both cases there is a dedicated OCM specific interpretation of content provided by the component model. But while the first solution allows to describe a closed scenario in a dedicated resource, where resources from dependent component version can be described by relative resource references and multiple scenarios can be separated by multiple flavors of this resource, the label-based approach is restricted to a local component version and a single scenario. Instead of an artifact type for the description, labels with a defined [name structure](./03-elements-sub.md#labels) are required.

## Artefact-Linking Label

This section defines a label convention for expressing cross-artefact relationships.
For example, it can be used to indicate that an SBoM resource describes a specific
OCI image resource within the same component version.

### Label Name

```

> ocm.software/artefact-references

Reviewer 2: nit: Do we think we will move towards a new spec version with `artifacts` instead of `resources` and `sources` soon? Currently, it seems rather like we will deprecate `sources` and keep the rest of the spec. So maybe `ocm.software/resource-references` would be more fitting?

Reviewer: The idea here was to make this label agnostic of `resources` and `sources` and allow usage on both of those `artifacts`. For this use case, the same terminology as in the [spec](https://github.com/open-component-model/ocm-spec/blob/main/doc/01-model/02-elements-toplevel.md#artifacts-resources-and-sources) was used.

```

> The label is a predefined label within the component model (see [Label Types](./07-extensions.md#label-types)).

Reviewer 2: ```suggestion
The label is a predefined label within the open component model (see [Label Types](./07-extensions.md#label-types)).
```

              foo: bar

  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: arm64
    labels:
      - name: ocm.software/artefact-references
        version: v1
        value:
          - identity:
              name: my-image
              version: 1.2.3
              foo: bar
```

### Lookup Algorithm

To find all artefacts related to a given subject resource:

1. Determine the identity of the subject resource: its `name`, `version`, and `extraIdentity`.
2. Iterate over all resources in the component descriptor.
3. For each resource, check whether it carries a label named `ocm.software/artefact-references` with `version: v1`.
4. If present, apply the following matching rules against each `identity` entry in the label value:
   - `name` MUST equal the subject's `name`.
   - If `version` is set, it MUST equal the subject's `version`.
   - The set of additional key-value pairs MUST exactly match the subject's `extraIdentity`:
     every key in the `identity` entry MUST be present and equal in the subject's `extraIdentity`,
     and the subject's `extraIdentity` MUST NOT contain any keys not present in the `identity` entry.

> 5. Resources that pass all checks are companions of the subject.

Reviewer 2: nit: the term "companion" is used nowhere else to describe this relationship, might be confusing

## Artefact-Linking Label

This section defines a label convention for expressing cross-artefact relationships.
For example, it can be used to indicate that an SBoM resource describes a specific
OCI image resource within the same component version.

### Label Name

```
odg.ocm.software/labels/artefact-ref/v1
```

The label follows the [vendor-specific label naming scheme](./07-extensions.md#label-types).
The version suffix (`v1`) is encoded in the label name; consumers MUST NOT treat a
label with a different version suffix as conforming to this convention.

### Placement

The label is placed on the **derived artefact**. The resource that is related to a
subject artefact carries the label pointing back to that subject. The subject
artefact itself requires no modification.

```
Component Descriptor
├── Resource: my-image          ← subject artefact, unchanged
└── Resource: my-image-sbom     ← derived artefact, carries the label
```

### Label Value

The label value is a YAML object with the following fields:

> **`artefactReference`** (required) - identifies the subject artefact within the same

Reviewer: To be consistent with existing concepts like the [srcRefs](https://ocm.software/docs/reference/component-descriptor/#component-resources-srcrefs-identityselector) property, I'd suggest to rename this field to `identitySelector` and only allow a flat mapping of identity relevant properties, e.g.:
```yaml
identitySelector:
  name: my-image
  version: 1.2.3
  architecture: amd64
```

component version:

- `name` (required) *string* — resource name of the subject artefact.
- `version` (optional) *string* — resource version. If omitted, any version matches.
- `extraIdentity` (optional) *map[string]string* — extra identity key-value pairs.
  Every entry listed here MUST be present and equal in the subject's `extraIdentity`.
  Required when multiple resources share the same name (e.g. arch-specific image variants).

**`metadata`** (optional) — additional context about the relationship:

- `relation` (optional) *string* — token describing the nature of the relationship.
  Recommended values:
  - `describes` — the derived artefact describes the subject (e.g. an SBoM).
  - `attests` — the derived artefact attests a property of the subject.

  Additional values MAY be defined by tooling. Consumers that do not recognise a
  value SHOULD treat the label as a generic artefact reference.

### Examples

An SBoM describing a single-variant image:

```yaml
resources:
  - name: my-image
    version: 1.2.3
    type: ociImage

  - name: my-image-sbom
    version: 1.2.3

>     type: application/spdx+json

Reviewer 3: From my understanding `type` would be `sbom`, and `application/spdx+json` would become the file's `mediaType`

See: https://ocm.software/docs/reference/input-and-access-types/#filev1

    labels:

>       - name: odg.ocm.software/labels/artefact-ref/v1

Reviewer 3: I would suggest to just name it `ocm.software/artefact-ref`

The linking will be part of OCM core, used by other tooling and not only ODG

There is also a dedicated `version` property for OCM labels, see: https://github.com/open-component-model/ocm-spec/blob/main/doc/01-model/03-elements-sub.md#labels

        value:
          artefactReference:
            name: my-image
          metadata:

>             relation: describes

Reviewer 4: The label name names the semantical convention we want to define as part of the ocm spec. It should not be odg specific and should contain the convention name. The version should be moved to the version field. The values then should focus on the actual reference logic. My idea:

```suggestion
    labels:
      - name: ocm.software/artefactReference
         version: v1
         value:
            identitySelector:
              name: my-image
              version: 1.2.3
              architecture: amd64
```

Is the metadata really needed? If the type of the resource which has the label is sbom, isn't that already descriptive enough?

Reviewer: > The label name names the semantical convention we want to define as part of the ocm spec.

Looking at the [spec](https://github.com/open-component-model/ocm-spec/blob/main/doc/01-model/07-extensions.md#label-types), shouldn't the label name then rather just be `artefactReference`?

> The version should be moved to the version field.

+1

> Is the metadata really needed?

I agree that we probably don't require the current `metadata` property. However, as also discussed in the hackathon, we will need some sort of mechanism to further specify the kind of reference (e.g. in case there are multiple SBOMs for the same artefact). I would suggest to use the `extraIdentity` field of the referencing artefact (here: the OCM resource for the SBoM) for this, e.g.:

```yaml
resources:
  - name: my-image
    version: 1.2.3
    type: ociImage
    extraIdentity:
      foo: bar
  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: amd64
    labels:
      - name: artefactReference
        version: v1
        value:
          identitySelector:
            name: my-image
            version: 1.2.3
            foo: bar
  - name: my-image-sbom
    version: 1.2.3
    type: sbom
    extraIdentity:
      architecture: arm64
    labels:
      - name: artefactReference
        version: v1
        value:
          identitySelector:
            name: my-image
            version: 1.2.3
            foo: bar  
```
Wdyt about this approach?

Reviewer 3: @8R0WNI3 I think that's really good

Just asking myself: if someone puts `name` and `version` into `extraIdentity`, what happens?
But having something like `identitySelector.extra` feels too nested :/

Reviewer: > If someone puts name and version into extraIdentity, what happens?

According to the [spec](https://github.com/open-component-model/ocm-spec/blob/main/doc/01-model/03-elements-sub.md#element-identity), the `name` **may not** be part of the extra identity. If the `version` is part of the extra identity, this does not matter (I think). To my knowledge, the component descriptors of Gardener even write the `version` _explicitly_ into the extra identity, thus making the implicit contract defined in the spec explicit.

> But having something like identitySelector.extra feels too nested.

I agree. Also, flattening the extra identity matches the behaviour implemented by the already existing [srcRefs](https://ocm.software/docs/reference/component-descriptor/#component-resources-srcrefs-identityselector) property.

```

Two derived artefacts referencing a specific architecture variant via `extraIdentity`:

```yaml
resources:
  - name: my-image
    version: 1.2.3
    type: ociImage
    extraIdentity:
      arch: amd64

  - name: my-image-sbom
    version: 1.2.3

>     type: application/spdx+json

Reviewer 4: The resource type should be `sbom`, see https://github.com/open-component-model/ocm-spec/blob/b286e28f290f33628a31def4689e1a9fb4fd6d4a/doc/04-extensions/01-artifact-types/sbom.md#L4

The media type of the resource access should be spdx+json
